### DISCUSSION TITLE
Configuring custom signature hash for TLS communication

### DISCUSSION THREAD

**Discussion by mohitvaid** [syslog-ng/syslog-ng#3983] [Q&A] [answered]
Hi there,

As part of our TLS configuration, to be compliant with [NIAP](https://www.niap-ccevs.org/product/)... we need to be specific wrt to few specific ciphers, elliptic curves and signature hash,
Based on https://www.syslog-ng.com/technical-documents/doc/syslog-ng-open-source-edition/3.24/administration-guide/61#TOPIC-1298105

I can see we can configure the ciphers and the elliptic curves but unable to find capability to configure the signature hash algorithms.

Any pointers to achieve the same?

**Comment by gaborznagy** [comment]
Maybe I'm missing something, but why can't you configure the hashing algorithm with `cipher-suite()` option?
Besides, what do you mean by "custom" hash?

I can't access the Slack chat you've referenced, maybe you could quote the essential parts here?

**Reply by mohitvaid** [reply]
@gaborznagy Thanks for looking at this discussion.

First of all want to clarify, by "custom" hash I meant to say that the desired signature hash, as configured/specified to be NIAP compliant.

As per NIAP requirement mentioned in https://commoncriteria.github.io/pp/tls/tls-release.pdf

Only SHA256, SHA384 & SHA512 are the only desired hash algorithms.

![image](https://user-images.githubusercontent.com/9480957/162960824-7dad90d3-6cde-4375-91d4-c02031acc2a7.png)

As of today we are using rsyslog, but moving to syslog-ng as we need to integrate the MQTT support.

And rsyslog provided the ability to configure:

- StreamDriverPermittedPeers
- StreamDriverAuthMode
- StreamDriver
- StreamDriverMode
- ClientSignatureAlgorithms
- Curves
- SignatureAlgorithms
- CipherString

Out of these, I can only find from https://www.syslog-ng.com/technical-documents/doc/syslog-ng-open-source-edition/3.24/administration-guide/61#TOPIC-1298105 that, syslog-ng supports configuring 

- peer-verify, 
- cipher-suite,
- ecdh-curve-list

My question is, is there any possible way to specify the desired SignatureAlgorithms and other missing configuration for the syslog-ng client configurations.

**Comment by gaborznagy** [comment]
Thanks for the details, it cleared out a lot.

I still think you can configure the signature algorithms through `cipher-suites()` options, as TLS cipher suites are containing the authentication algo, the MAC algo, etc.
[openssl ciphers documentation](https://www.openssl.org/docs/man1.1.1/man1/ciphers.html) helps with the syntax to define a stricter cipher suite.
I would recommend something `SHA256:SHA384:SHA512` or `!SHA1:etc.`.
You could explicitly disable the earlier TLS versions with `ssl-options(no-tlsv1,no-tlsv11)` too: https://www.syslog-ng.com/technical-documents/doc/syslog-ng-open-source-edition/3.36/administration-guide/73#TOPIC-1768778

About rsyslog options: some of those options are rsyslog-specific to configure its drivers (StreamDriverPermittedPeers, StreamDriverAuthMode, StreamDriver, StreamDriverMode).
I've checked their source code and they use [`SSL_CONF_cmd()`](https://www.openssl.org/docs/man1.1.1/man3/SSL_CONF_cmd.html) to set the signature algorithms to use. It is similar to [GnuTLS PriorityStrings](https://gnutls.org/manual/html_node/Priority-Strings.html).
Still, what can be configured through `SSL_CONF_cmd` can be through `cipher-suites()` too (which uses `SSL_CTX_set_cipher_list()` and `SSL_CTX_set_ciphersuites`).

**Reply by mohitvaid** [reply]
@gaborznagy I guess that is where the issue is with tls configuration in syslog-ng, 
To be precise, https://www.openssl.org/docs/man1.0.2/man3/SSL_CONF_cmd.html helps user configuring -sigalgs and -client_sigalgs but syslog-ng's use of SSL_CTX_set_cipher_list does not allow so.

FYI, i specified SHA256 in the like this cipher-suite("RSA-SHA256") and here is the "Client-Hello"


```Frame 76: 197 bytes on wire (1576 bits), 197 bytes captured (1576 bits) on interface \Device\NPF_{AB8F3132-60A4-4598-A16B-45ED10C5CEF3}, id 0
Ethernet II, Src: :e3:ff (00:50:56:88:e3:ff), Dst: :ec:4c (00:50:56:88:ec:4c)
Internet Protocol Version 4, Src: 10.108.120.149, Dst: 10.108.120.136
Transmission Control Protocol, Src Port: 14139, Dst Port: 49159, Seq: 1, Ack: 1, Len: 143
Transport Layer Security
    TLSv1 Record Layer: Handshake Protocol: Client Hello
        Content Type: Handshake (22)
        Version: TLS 1.0 (0x0301)
        Length: 138
        Handshake Protocol: Client Hello
            Handshake Type: Client Hello (1)
            Length: 134
            Version: TLS 1.2 (0x0303)
            Random: 97092c79c29aa8c6f1dacb7a0f4d6c36489e19c3b4ff8dcc4cc6f32404589bac
            Session ID Length: 0
            Cipher Suites Length: 28
            Cipher Suites (14 suites)
                Cipher Suite: TLS_DHE_RSA_WITH_AES_256_CBC_SHA256 (0x006b)
                Cipher Suite: TLS_DHE_DSS_WITH_AES_256_CBC_SHA256 (0x006a)
                Cipher Suite: TLS_DH_anon_WITH_AES_256_CBC_SHA256 (0x006d)
                Cipher Suite: TLS_RSA_WITH_AES_256_CBC_SHA256 (0x003d)
                Cipher Suite: TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256 (0xc027)
                Cipher Suite: TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256 (0xc023)
                Cipher Suite: TLS_DHE_RSA_WITH_AES_128_CBC_SHA256 (0x0067)
                Cipher Suite: TLS_DHE_DSS_WITH_AES_128_CBC_SHA256 (0x0040)
                Cipher Suite: TLS_DH_anon_WITH_AES_128_CBC_SHA256 (0x006c)
                Cipher Suite: TLS_ECDH_RSA_WITH_AES_128_CBC_SHA256 (0xc029)
                Cipher Suite: TLS_ECDH_ECDSA_WITH_AES_128_CBC_SHA256 (0xc025)
                Cipher Suite: TLS_RSA_WITH_AES_128_CBC_SHA256 (0x003c)
                Cipher Suite: TLS_RSA_WITH_NULL_SHA256 (0x003b)
                Cipher Suite: TLS_EMPTY_RENEGOTIATION_INFO_SCSV (0x00ff)
            Compression Methods Length: 1
            Compression Methods (1 method)
            Extensions Length: 65
            Extension: ec_point_formats (len=4)
            Extension: supported_groups (len=8)
                Type: supported_groups (10)
                Length: 8
                Supported Groups List Length: 6
                Supported Groups (3 groups)
                    Supported Group: secp256r1 (0x0017)
                    Supported Group: secp384r1 (0x0018)
                    Supported Group: secp521r1 (0x0019)
            Extension: session_ticket (len=0)
            Extension: signature_algorithms (len=32)
                Type: signature_algorithms (13)
                Length: 32
                Signature Hash Algorithms Length: 30
                Signature Hash Algorithms (15 algorithms)
                    Signature Algorithm: rsa_pkcs1_sha512 (0x0601)
                    Signature Algorithm: SHA512 DSA (0x0602)
                    Signature Algorithm: ecdsa_secp521r1_sha512 (0x0603)
                    Signature Algorithm: rsa_pkcs1_sha384 (0x0501)
                    Signature Algorithm: SHA384 DSA (0x0502)
                    Signature Algorithm: ecdsa_secp384r1_sha384 (0x0503)
                    Signature Algorithm: rsa_pkcs1_sha256 (0x0401)
                    Signature Algorithm: SHA256 DSA (0x0402)
                    Signature Algorithm: ecdsa_secp256r1_sha256 (0x0403)
                    Signature Algorithm: SHA224 RSA (0x0301)
                    Signature Algorithm: SHA224 DSA (0x0302)
                    Signature Algorithm: SHA224 ECDSA (0x0303)
                    Signature Algorithm: rsa_pkcs1_sha1 (0x0201)
                    Signature Algorithm: SHA1 DSA (0x0202)
                    Signature Algorithm: ecdsa_sha1 (0x0203)
            Extension: heartbeat (len=1)
```

But when using rsyslog, the client hello's signature sections looks like:

```
Extension: signature_algorithms (len=8)
    Type: signature_algorithms (13)
    Length: 8
    Signature Hash Algorithms Length: 6
    Signature Hash Algorithms (3 algorithms)
        Signature Algorithm: rsa_pkcs1_sha256 (0x0401)
        Signature Algorithm: rsa_pkcs1_sha384 (0x0501)
        Signature Algorithm: rsa_pkcs1_sha512 (0x0601)
```

Can we do something about this?

In short I want to restrict the **_Signature Hash Algorithms_**

**Comment by mohitvaid** [comment]
@gaborznagy 

Created https://github.com/syslog-ng/syslog-ng/issues/3993 as the feature request.

**Comment by mohitvaid** [comment]
@gaborznagy is there any timeline we can expect this adoption?

**Comment by MrAnno** [comment, marked as answer]
In summary, this question is about a TLS extension called `signature_algorithms`, which is specified in RFC5246 (for TLS 1.2) and RFC8446 (for TLS 1.3). It can be used to restrict which signature/hash pairs may be used in digital signatures.

syslog-ng supports specifying this extension field since v3.37.1, for example: `tls(sigalgs("RSA-PSS+SHA256:ed25519"))`
More info can be found in #4000.

**Reply by gaborznagy** [reply]
@mohitvaid #4000 is merged, so support for signature_algorithms will be released in the next version (3.37), which is expected to be around at the end of May.

In the meantime, you can try out the built rpm packages (https://github.com/syslog-ng/syslog-ng/actions/runs/2274810331), or create the packages yourself with `dbld/rules` (our docker based build system).

**Reply by mohitvaid** [reply]
@gaborznagy Can you please let us know the release timeline for 3.37? We are awaiting this feature.

**Reply by MrAnno** [reply]
v3.37 can be expected this Friday.
https://github.com/syslog-ng/syslog-ng/milestones

**Reply by mohitvaid** [reply]
@MrAnno I can not find the release build yet. Kindly update.

**Reply by MrAnno** [reply]
@mohitvaid 
https://github.com/syslog-ng/syslog-ng/releases/tag/syslog-ng-3.37.1
https://github.com/syslog-ng/syslog-ng#installation-from-binaries
